[mkdocs] docs: add cross chain swap tutorial and concept (#1925)
symbol/symbol: mkdocs/pages/en/textbook/cross-chain-swaps.md

# Cross-Chain Swaps

Cross-chain swap

> :   A cross-chain swap enables trading tokens (<mosaics:> in Symbol) across different blockchains without relying on an
>     intermediary (e.g., a centralized exchange).

Reviewer: We need to add the word "trustless".
The topic is complex, so we better keep it lean in this intro. We can add the link to <mosaic:> later on, when we give specific Symbol examples.

```suggestion
:   A cross-chain swap enables trading tokens trustlessly across different blockchains without relying on an
    intermediary (e.g., a centralized exchange).
```

> Since tokens cannot be transferred directly between blockchains with different technologies, the trade is performed
> inside each chain separately.
> The cross-chain swap protocol ensures that either both parties receive their funds or each party is refunded.

Reviewer: The part about funds and refunds is a bit confusing. How about:

```suggestion
Since tokens cannot be transferred directly between blockchains with different technologies,
a separate transfer is performed on each chain.
The cross-chain swap protocol ensures that either both transfers succeed or none does.
```

```dot
digraph CrossChainOverview {
    rankdir="LR";
    fontsize=12;

    subgraph clusterSwap {
        label="Cross-Chain Swap";

>         tooltip="Cross-Chain Swap";
>         subgraph clusterChainA {
>             label="Chain A";
>             tooltip="Chain A";
>             style=dashed;
>             AliceA [label="Alice" tooltip="Alice"];
>             BobA [label="Bob" tooltip="Bob"];
>             AliceA -> BobA [label="Token A"];
>         }
>         subgraph clusterChainB {
>             label="Chain B";
>             tooltip="Chain B";
>             style=dashed;
>             AliceB [label="Alice" tooltip="Alice"];
>             BobB [label="Bob" tooltip="Bob"];
>             AliceB -> BobB [label="Token B" dir=back];
>         }
>     }
> }

Reviewer: This diagram looks exactly like the one for aggregates, so it's confusing.
I think we can make it clearer that these are separate chains by not grouping them.

```suggestion
digraph CrossChainOverview {
    rankdir="LR";
    fontsize=12;

    subgraph clusterChainB {
        label="Chain B";
        tooltip="Chain B";
        AliceB [label="Alice" tooltip="Alice"];
        BobB [label="Bob" tooltip="Bob"];
        AliceB -> BobB [label="Token B" dir=back];
    }
    subgraph clusterChainA {
        label="Chain A";
        tooltip="Chain A";
        AliceA [label="Alice" tooltip="Alice"];
        BobA [label="Bob" tooltip="Bob"];
        AliceA -> BobA [label="Token A"];
    }
}
```

```

This is conceptually similar to an <aggregate transaction:>, but instead of grouping operations within one blockchain,
the swap coordinates operations across two independent chains.

## Protocol

> Symbol follows the [Hashed TimeLock Contract](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts) (HTLC) protocol

Reviewer: Hmmm... it's not that Symbol follows this protocol. It's more like Symbol offers functionality (hashlocks and timelocks) that allows you to implement this protocol, no?

to create a trustless environment for the decentralized exchange of assets.

> HTLC uses two mechanisms to eliminate counterparty risk:

Reviewer: "Counterparty risk" sounds weird. Risk of what?

How about:
```suggestion
HTLC uses two mechanisms to provide trustless swaps:
```

Author: I've simplified the term "counterpart" for "the other party" or "the other chain".

> 

Reviewer: Worth creating a glossary entry for HTLC. If not here, maybe in the Glossary page with a link back here.

Reviewer: Now that I see it in the glossary, I think it looks better in this same page, so the content is not spread out.
I'll take care of it.

Hashlock
:   A cryptographic hash of a secret value called the _proof_.
    Funds can only be claimed by revealing the proof that produces this hash.

>     Revealing the proof on one chain allows the counterparty to read it and use it on the other chain.

Reviewer: The first two sentences are OK, but the third one is specific to the cross-chain swap use case, which is not the only use case for hashlocks (I guess).

So I would take the third sentence out of the definition box, and introduce it with something like "For example, for the cross-chain swap use case, ...".

Timelock

> :   A deadline after which unclaimed funds are returned to their original owner.

Reviewer: A timelock is a "deadline"? Or is it a kind of transaction in which one party locks funds until a deadline expires or a proof is revealed?
Are hashlocks and timelocks related? Or only related in the cross-chain swap use case? I am confused :D

Author: Changed to “a condition” to clearly differentiate that it is a concept from the protocol rather than a deadline or a specific transaction in Symbol. Then, restructured the section to make clear what's the protocol, what's specific to symbol and what's specific to other chain.

    In Symbol, the maximum secret lock duration is 365 days.

> In Symbol, a single <ser:SecretLockTransactionV1> locks funds by specifying both the hashlock (`secret` field) and the

Reviewer: This is the Textbook, meant for all audiences, so it cannot have links to the Devbook (leave those for the dev tutorials). You have to explain stuff in a more general way.
You can say something like "Symbol offers a transaction type called SecretLock which..."

timelock (`duration` field), along with the hash algorithm and the recipient.
A <ser:SecretProofTransactionV1> unlocks them by revealing the proof that matches the hashlock.

The key insight is that the **same hashlock** is used on both blockchains.

> When one party reveals the proof on one chain, the other party can read it and use it on the other chain.

Reviewer: It is true that this is the key insight, so it needs to be extremely clear.
The passive voice makes it hard to know who is doing each action.

Maybe do something like: a party initiates the transfer on one chain but locks it with a hashlock. The transfer will not be completed until the secret is revealed. The second party then announces its transfer using the same hashlock, so in order for the first party to receive it it must publish the secret. This in turn allows the second party to unlock the initial transfer and complete the swap.

The timelocks must be configured so that neither party can cheat the other

> (see [Timing Constraints](#timing-constraints)).

Reviewer: ```suggestion
(see [Timing Constraints](#timing-constraints) below).
```

So readers know that they don't need to click now and stop reading, because it will be explained later.

## Example

Alice and Bob want to exchange tokens across two different blockchains, Chain A and Chain B.

> Both Alice and Bob have accounts on both chains.

Reviewer: You need to give some more details of what you are about to do so the steps are easier to understand.
So, on which chain Alice is _sending_ and on which chain Alice is _receiving_?
If it's the same case as in the diagram in the intro, then we have to say so.

The swap proceeds in four steps:

```graphviz neato timelock-windows.svg
digraph TimelockWindows {
    node [fontsize=14];

>     pad=0.5;
> 
>     // Alice's lock bar (Chain A - longer, top)
>     A_start [label="" shape=point pos="0,5!" width=0.1];
>     A_end [label="" shape=point pos="12,5!" width=0.1];
>     A_start -> A_end [style=bold penwidth=3 arrowhead=none];
>     A_label [label="Alice's lock on Chain A (longer timelock)"
>         shape=plaintext pos="6,5.3!" fontsize=14];
> 
>     // Timelock Difference (between Bob's expiry and Alice's expiry)
>     TG_start [label="" shape=point pos="7,3.5!" width=0.1];
>     TG_end [label="" shape=point pos="12,3.5!" width=0.1];
>     TG_start -> TG_end [style=dashed penwidth=2 arrowhead=none];
>     TG_label [label="Timelock Difference"
>         shape=plaintext pos="9.5,3.8!" fontsize=14];
> 
>     // Bob's lock bar (Chain B - shorter)
>     B_start [label="" shape=point pos="1,2!" width=0.1];
>     B_end [label="" shape=point pos="7,2!" width=0.1];
>     B_start -> B_end [style=bold penwidth=3 arrowhead=none];
>     B_label [label="Bob's lock on Chain B (shorter timelock)"
>         shape=plaintext pos="4,2.3!" fontsize=14];
> 
>     // Vertical ticks
>     T1_top [label="" shape=point pos="0,5!" width=0];
>     T1_bot [label="" shape=point pos="0,0.8!" width=0];
>     T1_top -> T1_bot [arrowhead=none style=dotted];
> 
>     T2_top [label="" shape=point pos="1,2!" width=0];
>     T2_bot [label="" shape=point pos="1,0.8!" width=0];
>     T2_top -> T2_bot [arrowhead=none style=dotted];
> 
>     T3_top [label="" shape=point pos="6,2!" width=0];
>     T3_bot [label="" shape=point pos="6,0.8!" width=0];
>     T3_top -> T3_bot [arrowhead=none style=dotted];
> 
>     T4_top [label="" shape=point pos="7,2!" width=0];
>     T4_bot [label="" shape=point pos="7,0.8!" width=0];
>     T4_top -> T4_bot [arrowhead=none style=dotted];
> 
>     T5_top [label="" shape=point pos="10,5!" width=0];
>     T5_bot [label="" shape=point pos="10,0.8!" width=0];
>     T5_top -> T5_bot [arrowhead=none style=dotted];
> 
>     T6_top [label="" shape=point pos="12,5!" width=0];
>     T6_bot [label="" shape=point pos="12,0.8!" width=0];
>     T6_top -> T6_bot [arrowhead=none style=dotted];
> 
>     // Labels below
>     L1 [label="1. Alice\nlocks" shape=plaintext pos="0,0.2!" fontsize=13];
>     L2 [label="2. Bob\nlocks" shape=plaintext pos="1,0.2!" fontsize=13];
>     L3 [label="3. Alice\nreveals proof" shape=plaintext pos="6,0.2!" fontsize=13];
>     L4 [label="Bob's timelock\nexpires" shape=plaintext pos="7,0.2!" fontsize=13];
>     L5 [label="4. Bob\nwithdraws" shape=plaintext pos="10,0.2!" fontsize=13];
>     L6 [label="Alice's timelock\nexpires" shape=plaintext pos="12,0.2!" fontsize=13];
> }

Reviewer: The graph is a bit difficult to read so I made everything a bit bigger.
I also could not help optimizing the code to make it easier to maintain 😅

```suggestion
digraph TimelockWindows {
    node [fontsize=24 shape=point width=0];
    edge [arrowhead=none];
    pad=0.5;

    // Alice's lock bar (Chain A - longer, top)
    A_start [pos="0,5!" width=0.2];
    A_end [pos="12,5!" width=0.2];
    A_start -> A_end [style=bold penwidth=5];
    A_label [label="Alice's lock on Chain A (longer timelock)" shape=plaintext pos="6,5.3!"];

    // Timelock Difference (between Bob's expiry and Alice's expiry)
    TG_start [pos="7,3.5!" width=0.2];
    TG_end [pos="12,3.5!" width=0.2];
    TG_start -> TG_end [style=dashed penwidth=5];
    TG_label [label="Timelock Difference" shape=plaintext pos="9.5,3.8!"];

    // Bob's lock bar (Chain B - shorter)
    B_start [pos="1,2!" width=0.2];
    B_end [pos="7,2!" width=0.2];
    B_start -> B_end [style=bold penwidth=5];
    B_label [label="Bob's lock on Chain B (shorter timelock)" shape=plaintext pos="4,2.3!"];

    // Vertical ticks
    T1_top [pos="0,5!"];
    T1_bot [pos="0,0.8!"];
    T1_top -> T1_bot [style=dotted];

    T2_top [pos="1,2!"];
    T2_bot [pos="1,0.8!"];
    T2_top -> T2_bot [style=dotted];

    T3_top [pos="5,2!"];
    T3_bot [pos="5,0.8!"];
    T3_top -> T3_bot [style=dotted];

    T4_top [pos="7,2!"];
    T4_bot [pos="7,0.8!"];
    T4_top -> T4_bot [style=dotted];

    T5_top [pos="10,5!"];
    T5_bot [pos="10,0.8!"];
    T5_top -> T5_bot [style=dotted];

    T6_top [pos="12,5!"];
    T6_bot [pos="12,0.8!"];
    T6_top -> T6_bot [style=dotted];

    // Labels below
    L1 [label="1.\nAlice\nlocks" shape=plaintext pos="0,0.2!"];
    L2 [label="2.\nBob\nlocks" shape=plaintext pos="1,0.2!"];
    L3 [label="3. Alice\nclaims and\nreveals proof" shape=plaintext pos="5,0.2!"];
    L4 [label="Bob's\ntimelock\nexpires" shape=plaintext pos="7,0.2!"];
    L5 [label="4. Bob\nclaims\n " shape=plaintext pos="10,0.2!"];
    L6 [label="Alice's\ntimelock\nexpires" shape=plaintext pos="12,0.2!"];
}
```

```

1. **Alice locks tokens on Chain A:** Alice generates a random proof and computes its <hashlock:>.

>     She creates a lock on Chain A using this hashlock, naming Bob as the recipient and setting an expiration

Reviewer: It is worth mentioning the word "transfer", maybe repeating parts of the explanation of the protocol.

>     (<timelock:>).

Reviewer: No need for the parenthesis, no?

2. **Bob locks tokens on Chain B:** Bob reads the hashlock from Alice's lock on Chain A and creates his own lock using
    the **same hashlock** on Chain B, naming Alice as the recipient.

>     Bob's timelock must expire **before** Alice's timelock on Chain A.

Reviewer: What happens if not? The problem is that then Alice can wait until Bob's timelock expires, she gets her money back in chain A, and then announces the proof on chain B and gets Bob's money too.

This has to be explained, probably in an info box inside step 2, because it's a long explanation.

3. **Alice claims on Chain B:** Alice reveals the proof on Chain B to claim Bob's locked tokens.
4. **Bob claims on Chain A:** Bob reads the proof from Chain B and uses it to claim Alice's locked tokens on Chain A.

> The protocol ensures that, if the timelocks are configured correctly, either both parties receive their funds or each

Reviewer: Well, if Bob does not claim in time then he loses his money, so the protocol doesn't "ensure" that everything works.

You can say instead that the protocol ensures every participant has a fair chance to receive their part of the swap.

party is refunded:

> | Scenario                        | Outcome                                                                                                              |

Reviewer: ```suggestion
| Scenario                                        | Outcome                                                                                              |
```

| ------------------------------------------------| -----------------------------------------------------------------------------------------------------|
| Bob does not lock tokens                        | Alice's tokens are refunded after her timelock on Chain A expires.                                   |
| Alice does not reveal proof                     | Bob's tokens are refunded when his timelock expires. Alice's tokens are also refunded later.         |
| Alice reveals proof, Bob claims in time         | Alice gets Bob's tokens. Bob can claim Alice's tokens before her timelock expires.                   |
| Alice reveals proof, Bob does not claim in time | **Worst case for Bob**. Alice keeps Bob's tokens and gets her own back when her timelock expires.    |

See [Timing Constraints](#timing-constraints) for details on how to configure the timelocks correctly.

## Timing Constraints

The safety of the swap depends on how the timelocks are configured relative to each other.
The following sections use the Alice and Bob [example](#example) for clarity.

> ### Timelock Ordering

Reviewer: This is the explanation that I was missing before. Then maybe we don't need an info box in the example and just link to this section for more details.

Since Alice knows the proof from the start, she controls when to reveal it.
Bob's timelock must expire **before** Alice's timelock to prevent Alice from waiting for her own refund and then still
claiming Bob's tokens by revealing the proof.

### Timelock Difference

The difference between Bob's timelock expiry and Alice's timelock expiry must be **large enough** for Bob to detect the
proof on Chain B, submit his claim on Chain A, and get it confirmed.

In the worst case, Alice reveals the proof at the last moment before Bob's lock expires, leaving Bob only this

> difference to react.

Reviewer: Worth mentioning that if Bob thinks the window that Alice provided is not big enough, he can simply refuse to perform the swap and request a bigger window.

The difference should account for:

> * **Finality periods** on both chains.

Reviewer: Add a link to the Finality section below, because, on its own, this is not clear.

* **Observation time** for Bob to detect the proof on Chain B.
* **Transaction submission and confirmation time** on Chain A.

> * **Possible resubmission** if Bob's claim transaction is rolled back.

Reviewer: "Rolled back" makes me think of chain reorgs, but this is not what you mean, right?
You mean that the transaction is not accepted (for example for too low fees) and Bob needs to resubmit?

See the [timing diagram](#example) in the example above.

### Finality

> Each party must wait for the previous step's transaction to reach [finality](consensus.md#finalization)

Reviewer: Better link to `<finalization:>`.

on their chain before acting on the next step.

In particular, Alice's lock (step 1) must be final before Bob locks, and Bob's lock (step 2) must be final before
Alice reveals the proof.
A <rollback:> at either point could remove a lock after the counterparty has already acted.

If a claim transaction (steps 3 or 4) is rolled back, the party can resubmit as long as the timelock has not expired.

> ### Front-Running Risk

Reviewer: Nice section!

When Alice reveals the proof on Chain B (step 3), the proof may be visible in the <unconfirmed pool:> before the
transaction is confirmed.
An observer on Chain B could extract the proof and attempt to claim funds on either chain before the intended
recipient does.

In Symbol, anyone can submit the proof, but the funds are always sent to the recipient specified in the lock, which
eliminates this risk.
The counterpart chain's HTLC implementation must also enforce recipient-only claiming to prevent this attack.
Verify this before using any HTLC contract for a cross-chain swap.

Even if Alice's proof transaction fails to confirm, the proof is already exposed.
This does not break the protocol: Alice can resubmit on Chain B, and Bob can use the revealed proof on Chain A.
Both parties can still claim their funds as long as their respective timelocks have not expired.

## Supported Chains

Symbol's secret lock mechanism can be used with any blockchain that supports HTLC or an equivalent locking mechanism
with a compatible hash algorithm.

> The counterpart chain must support locking funds with one of the hash algorithms listed in <ser:LockHashAlgorithm>.

Reviewer: No links to the devbook allowed :)

There are only 3 algorithms, so you can copy the whole table here.

each party waits for the right confirmations, and both parties understand how the public proof can be used against
them.

The following sections cover each of these concerns, using the Alice and Bob [example](#example) for clarity.

### Lock Verification

Before acting on the other party's lock, each party must verify that the lock actually matches what was agreed.
A mechanical follow of the protocol without this check allows the other side to cheat by posting a lock with the
wrong amount, token, hashlock, or recipient.

Each party should check:

* **Amount and token:** the lock holds the expected quantity of the expected token.
* **Hashlock:** the lock uses the agreed hashlock (same value on both chains).
* **Recipient:** the lock names the correct recipient.
* **Timelock:** the expiry is within the agreed ordering (see [Timelock Ordering](#timelock-ordering) and
    [Timelock Difference](#timelock-difference)).

Concretely:

* **Before Bob locks** (step 2), Bob verifies Alice's lock on Chain A.
* **Before Alice reveals the proof** (step 3), Alice verifies Bob's lock on Chain B.

If any check fails, the party should refuse to continue.

> Alice can wait for her own timelock to expire and reclaim her tokens; Bob has not yet locked.

Author: Added this subsection since the latest review

claiming Bob's tokens by revealing the proof.

### Timelock Difference

The difference between Bob's timelock expiry and Alice's timelock expiry must be **large enough** for Bob to detect the
proof on Chain B, submit his claim on Chain A, and get it confirmed.

In the worst case, Alice reveals the proof at the last moment before Bob's lock expires, leaving Bob only this
difference to react.

The difference should account for:

* **Finality periods** on both chains (see [Finality](#finality) below).
* **Observation time** for Bob to detect the proof on Chain B.
* **Transaction submission and confirmation time** on Chain A.
* **Possible resubmission** if Bob's claim transaction is not accepted by the network.

If Bob considers the window Alice provided too small, he can refuse to perform the swap and ask Alice to set a longer
timelock on Chain A.

See the [timing diagram](#example) in the example above.

### Late Proof Revelation

Alice must reveal the proof well before Bob's timelock expires.
If she reveals it close to that deadline and her claim transaction on Chain B does not confirm in time, Bob gets
refunded on Chain B while the proof is now public.

> Bob can then use the proof to claim Alice's tokens on Chain A, leaving Alice with nothing.

Author: Added this subsection since the latest review
